### PR TITLE
Enable index scan on AO tables for indexes that don’t have amgetbitmap

### PR DESCRIPTION
**Pull request by fanfuxiaoran**
Ivfflat and hsnw indexes both don't support 'amgetbitmap', after this pr they can be used on AO tables.

The main reason for us to enable those 2 indexes on AO tables is that they both only support `ORDER BY` queries and the typical query on embedding data is `ORDER BY embedding LIMIT `, as only limited rows fetched from the base tables, the cost on AO table random access won't be big.

The other reason is that the IVFFLAT index only fetches part of the index tuples when doing a query (the percent is probes/lists), so generally using the index can improve the performance.

No such indexes (amgetbitmap=NULL but amgettuple=non-NULL) exist in core or have ever been supported for GPDB customers, so no chance of regression with this change.

No test has been added to this pr because there isn't any index that satisfies the condition (amgetbitmap=NULL but amgettuple=non-NULL),  it can only be tested in other places. And the behaviors of the indexes that have existed in the core don't change, so no new test cases are needed.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/optimizer/path/indxpath.c [hunk 1]
```diff
@@ -809,7 +809,8 @@ get_index_paths(PlannerInfo *root, RelOptInfo *rel,
 		if (index->amhasgettuple &&
 				((!IsAccessMethodAO(rel->relam) ||
 				 index->amcostestimate == bmcostestimate ||
-				 ipath->path.pathtype == T_IndexOnlyScan)))
+				 ipath->path.pathtype == T_IndexOnlyScan ||
+				 (!index->amhasgetbitmap))))
```
Comment (kainwen) [inline src/backend/optimizer/path/indxpath.c:813]: Better to add some simple comments here on why the change and also mention why we not add case here. (Not easy to introduce such AM and there are extensions that can benefit here.

Comment (soumyadeep2007) [reply src/backend/optimizer/path/indxpath.c:816]: Here's a full refactor of the comment:

```
/*
 * Random access to Append-Only is slow because AO doesn't use the buffer
 * pool and we want to avoid decompressing blocks multiple times.
 *
 * Bitmap scans on the other hand are processed in TID order and the
 * AO table AMs optimize fetches in TID order by keeping the last
 * decompressed block between fetch calls.
 *
 * Thus, we ban index scans on append-optimized tables, and generally
 * only pick bitmap index scans.
 *
 * Exceptions:
 * (1) Index-only scans as they don't need to access the base table.
 * (2) If the index AM does not support a bitmap index scan, like
 * pgvector's ivfflat or hsnw.
 */
```

**Review by soumyadeep2007 — APPROVED** [review]
LGTM!
One quick suggestion for the commit message is to show a simple pgvector ivfflat EXPLAIN ANALYZE with this patch applied on an AO/CO table, as motivation.

**Comment by fanfuxiaoran** [comment]
Hi, I did some tests to compare the performance between using seq scan and  ivfflat index scan on my laptop, the main two factors that impact the IVFFLAT index performance are `probe` and the number of `limit by`.  (pgvector reference https://github.com/pgvector/pgvector)

```
CREATE TABLE millions  WITH (appendoptimized=true)
AS (SELECT generate_series(1, 1000000) id)
DISTRIBUTED BY (id);

CREATE OR REPLACE FUNCTION generate_random_embedding (size integer, min float8, max float8)
RETURNS vector
AS $$
  import numpy as np
  return np.random.uniform(low=min, high=max, size=(size,)).tolist()
$$ LANGUAGE plpython3u;

CREATE TABLE embeddings_ao_1 
WITH (appendoptimized=true) AS 
(SELECT id, generate_random_embedding(512, -7.99, 3.15) embedding FROM ten_millions)
DISTRIBUTED BY (id);

ALTER TABLE embeddings_ao_1 
ALTER COLUMN embedding TYPE vector(512);

CREATE TABLE embeddings_ao_2
WITH (appendoptimized=true) AS 
(SELECT id, generate_random_embedding(512, -7.99, 3.15) embedding FROM ten_millions)
DISTRIBUTED BY (id);

ALTER TABLE embeddings_ao_2
ALTER COLUMN embedding TYPE vector(512);

-- CREATE IVFFLAT index on 
CREATE INDEX ON embeddings_ao_2
USING ivfflat (embedding vector_l2_ops) 
WITH (lists = 1000);
```
Create AO tables 'embeddings_ao_1' and e'mbeddings_ao_2', then inserted 1000000 rows, and only created IVFFLAT index on 'embeddings_ao_2' with 1000 lists, which means the data will be clustered in 1000 centers in the index.

I executed the following two queries above the 2 tables, as IVFFLAT only can be used on `ORDER BY` queries, and `order by limit xx` is the typical query used for embedding.
```
SELECT *
FROM embeddings_ao_*
ORDER BY embedding <-> (select generate_random_embedding(512, -7.99, 3.15)) ;

SELECT *
FROM embeddings_ao_*
ORDER BY embedding <-> (select generate_random_embedding(512, -7.99, 3.15)) limit 10;
```

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head></head><body>

  | order by embedding (limit 1000000) | order by embedding limit 10 | Probe
-- | -- | -- | --
embeddings_ao_1 | 9684.248  ms | 661.709 ms |  
embeddings_ao_2 | 653.436 ms | 348.631 ms | 1
embeddings_ao_2 | 588.257 ms | 354.173 ms | 10
embeddings_ao_2 | 4379.529 | 437.772 ms | 100
embeddings_ao_2 | 17190.770 ms ( planner choose seqscan) | 577.755  (planner choose seqscan) | 1000

</body></html>


The plan is as below when probes is '100'
```
explain analyze SELECT *
FROM embeddings_ao_2
ORDER BY embedding <-> (select generate_random_embedding(512, -7.99, 3.15)) limit 10;
                                                                                  QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=396.07..396.22 rows=10 width=44) (actual time=21.101..21.653 rows=10 loops=1)
   InitPlan 1 (returns $0)  (slice2)
     ->  Result  (cost=0.00..0.26 rows=1 width=32) (actual time=305.396..305.397 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=396.07..396.53 rows=30 width=44) (actual time=21.099..21.638 rows=10 loops=1)
         Merge Key: ((embeddings_ao_2.embedding <-> $0))
         ->  Limit  (cost=396.07..396.13 rows=10 width=44) (actual time=11.788..13.337 rows=10 loops=1)
               ->  Index Scan using embeddings_ao_2_embedding_idx on embeddings_ao_2  (cost=180.00..2340.67 rows=333334 width=44) (actual time=11.787..13.334 rows=10 loops=1)
                     Order By: (embedding <-> $0)
 Optimizer: Postgres query optimizer
 Planning Time: 29.715 ms
   (slice0)    Executor memory: 75K bytes.
   (slice1)    Executor memory: 329K bytes avg x 3 workers, 348K bytes max (seg0).
   (slice2)    Executor memory: 40K bytes.
 Memory used:  128000kB
 Execution Time: 334.898 ms
```
`ORDER BY XX LIMIT YY` is the typical query used on embedding, thus enable `IVFFLAT` on AO tables can significantly improve the performance.

**Comment by soumyadeep2007** [comment]
Thanks for sharing the results! Can you also show how long seq scan takes for these queries (if the indexes weren't picked) and add that to your table?

Also what is meant by `probe` in this context?

PS: If you want to know exactly how many times the index was asked to fetch a tuple, you can insert a elog in `appendonly_index_fetch_tuple`

**Comment by fanfuxiaoran** [comment]
> Thanks for sharing the results! Can you also show how long seq scan takes for these queries (if the indexes weren't picked) and add that to your table?

The table `embeddings_ao_1` is the same as the table `embeddings_ao_2`, but there isn't any index on it. 

> Also what is meant by `probe` in this context?

`probes` is a GUC of ivfflat index
```
When querying, specify an appropriate number of [probes](#query-options) (higher is better for recall, lower is better for speed) - a good place to start   is `sqrt(lists)`
```
The ivfflat index clusters the data into `lists` data centers, when querying, it only scans `probes` number of data centers and computes the distance, then order the data by distance. For example, if the `lists` is 1000 and `probes` is 10, when executing the query `SELECT * FROM embeddings_ao_2 ORDER BY embedding <-> (select generate_random_embedding(512, -7.99, 3.15)) limit 10;`, it firstly picks up 10 top centers which are close to vector `generate_random_embedding(512, -7.99, 3.15)`, then only scans the 10 data centers and return to client.

**Review by haolinw — APPROVED** [review]
